src/dags/dag.py

```diff
@@ -45,19 +45,94 @@ def concatenate_functions(
+    """Build a directed acyclic graph (DAG) from functions.
+
+    Functions can depend on the output of other functions as inputs, as long as the
+    dependencies can be described by a directed acyclic graph (DAG).
+
+    Functions that are not required to produce the targets will simply be ignored.
+
+    Args:
+        functions (dict or list): Dict or list of functions. If a list, the function
+            name is inferred from the __name__ attribute of the entries. If a dict,
+            the name of the function is set to the dictionary key.
+        targets (str): Name of the function that produces the target or list of such
+            function names.
+
+    Returns:
+        dag: the DAG (as networkx.DiGraph object)
+
+    """
+    # Harmonize and check arguments.
+    _functions, _targets = harmonize_and_check_functions_targets(functions, targets)
+
+    # Create the DAG
     _raw_dag = _create_complete_dag(_functions)
-    _dag = _limit_dag_to_targets_and_their_ancestors(_raw_dag, _targets)
-    _arglist = _create_arguments_of_concatenated_function(_functions, _dag)
-    _exec_info = _create_execution_info(_functions, _dag)
+    dag = _limit_dag_to_targets_and_their_ancestors(_raw_dag, _targets)
+
+    # Check if there are cycles in the DAG
+    _fail_if_dag_contains_cycle(dag)
+
+    return dag
+
+
+def create_combined_function_from_dag(
+    dag,
+    functions,
+    targets,
+    return_type="tuple",
+    aggregator=None,
+    enforce_signature=True,
+):
+    """Create combined function which allows to execute a complete directed acyclic
+    graph (DAG) in one function call.
+
+    The arguments of the combined function are all arguments of relevant functions that
+    are not themselves function names, in alphabetical order.
+
+    Args:
+        dag (networkx.DiGraph): a DAG of functions
+        functions (dict or list): Dict or list of functions. If a list, the function
+            name is inferred from the __name__ attribute of the entries. If a dict, the
+            name of the function is set to the dictionary key.
+        targets (str): Name of the function that produces the target or list of such
+            function names.
+        return_type (str): One of "tuple", "list", "dict". This is ignored if the
+            targets are a single string or if an aggregator is provided.
+        aggregator (callable or None): Binary reduction function that is used to
+            aggregate the targets into a single target.
+        enforce_signature (bool): If True, the signature of the concatenated function
+            is enforced. Otherwise it is only provided for introspection purposes.
+            Enforcing the signature has a small runtime overhead.
+
+    Returns:
+        function: A function that produces targets when called with suitable arguments.
+
+    """
+
+    # Harmonize and check arguments.
+    _functions, _targets = harmonize_and_check_functions_targets(functions, targets)
```

Comment:
Once this function is private, we can change it such that functions and targets already need to be harmonized and we skip the duplicate processing.

Reply:
But how would be make sure that functions and targets are harmonized when `_create_combined_function_from_dag` is called in `concatenate_functions`?

Two possibilities:
- Let `create_dag` return harmonized functions and targets (does not sound ideal)
- Call `_harmonize_and_check_functions_and_targets` in `concatenate_functions` (but still need to call it in `create_dag` for users only calling this function). So we would not get rid of duplicate processing.

Reply:
Also, the output is different for targets consisting of a string and targets consisting of a one-element list. If we would like to keep this behaviour, I think it is the easiest to keep `harmonize_and_check_functions_targets` here.

Reply:
I see that you need to do the harmonization twice because you do not return the intermediate outputs from `create_dags`. Getting rid of the duplicated call would need more restructuring. Since the processing function should be very fast to evaluate, we can leave it for now. Thanks for the clarification.